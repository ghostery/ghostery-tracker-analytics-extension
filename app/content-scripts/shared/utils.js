/* CSS STYLES HELPER FUNCTION */
/* ------------------------------------------------------------------------------------- */
export const assignStyles = (element, styles) => {
  let styleString = '';
  Object.keys(styles).forEach((styleKey) => {
    styleString += `${styleKey}: ${styles[styleKey]} !important;`;
  });
  element.setAttribute('style', styleString);
};

/* ELEMENT CREATOR HELPER FUNCTION */
const _createEl = (details) => {
  const { type, styles, src, alt, innerText } = details;

  const el = document.createElement(type);

  if (styles) assignStyles(el, styles);
  if (src) el.src = [chrome.extension.getURL(`${src}`)];
  if (alt) el.alt = alt;
  if (innerText) el.innerText = innerText;

  return el;
};

/* DIV ELEMENT CREATOR HELPER FUNCTION */
/* ------------------------------------------------------------------------------------- */
export const createDiv = (styles = {}, innerText = '') => {
  if (typeof styles === 'string') { // only an innerText param was passed
    return _createEl({ type: 'div', innerText: styles });
  }
  return _createEl({ type: 'div', styles, innerText });
};

/* IMG ELEMENT CREATOR HELPER FUNCTION */
/* ------------------------------------------------------------------------------------- */
export const createImg = (styles, src, alt = 'Decorative image') => _createEl({ type: 'img', styles, src, alt });

/* LINK ELEMENT CREATOR HELPER FUNCTION */
/* ------------------------------------------------------------------------------------- */
export const createLink = (styles, innerText) => _createEl({ type: 'a', styles, innerText });

/* SPAN ELEMENT CREATOR HELPER FUNCTION */
/* ------------------------------------------------------------------------------------- */
export const createSpan = (styles, innerText) => _createEl({ type: 'span', styles, innerText });

/* ADD OPEN SANS FONT TO THE DOCUMENT */
/* ------------------------------------------------------------------------------------- */
export const loadFonts = () => {
  // If we need to inject any other Open Sans fonts (eg. Italic, Light, etc.),
  // we can add them into this object using the appropriate parameters
  // (see /app/utils/scss/open-sans.scss for reference)
  const fontsToLoad = [
    { boldness: 'Regular', style: 'normal', weight: 400 },
    { boldness: 'SemiBold', style: 'normal', weight: 600 },
    { boldness: 'Bold', style: 'normal', weight: 700 },
  ];

  fontsToLoad.forEach((fontObject) => {
    const { boldness, style, weight } = fontObject;
    const openSansFont = new FontFace(
      'OpenSans',
      `url(${[chrome.extension.getURL(`dist/fonts/open-sans/OpenSans-${boldness}.ttf`)]})`,
      { style, weight },
    );
    openSansFont.load().then(loadedFont => document.fonts.add(loadedFont)).catch();
  });
};
