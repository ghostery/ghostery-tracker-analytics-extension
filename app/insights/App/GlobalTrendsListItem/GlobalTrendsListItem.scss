@import '../../../utils/scss/color-variables.scss';
@import '../../../utils/scss/category-background-colors.scss';

.GlobalTrendsListItem:last-of-type {
  .GlobalTrendsListItem__header {
    margin-bottom: 0;
  }
}

.GlobalTrendsListItem__header {
  margin-bottom: 3px;
  background-color: $black;
  padding: 10px;
  cursor: pointer;

  p {
    margin: 0 20px;
  }
}

.GlobalTrendsListItem__circle {
  margin-right: 9px;
  border-radius: 11px;
  width: 11px;
  height: 11px;
}

.GlobalTrendsListItem__star {
  width: 13px;
  height: 13px;
  cursor: pointer;
}

.GlobalTrendsListItem__trackingType {
  margin-top: 4px;
  margin-bottom: 10px;
}

.GlobalTrendsListItem__notDetectedText {
  margin: 5px 0;
  color: $ghosty-blue;
}

.GlobalTrendsListItem__trackingTypeText {
  margin: 0 12px 0 15px;
  font-size: 13px;
  color: $ghosty-blue;
}

.GlobalTrendsListItem--largeFont {
  margin-top: 5px;
  margin-bottom: 0;
  font-size: 20px;
  color: $ghosty-blue;
}

.GlobalTrendsListItem--smallFont {
  margin-bottom: 0;
  font-size: 10px;
}

.GlobalTrendsListItem--paddingZero {
  padding: 0;
}

.GlobalTrendsListItem--paddingZeroTop {
  padding: 1px 0 0;
}

.GlobalTrendsListItem--paddingZeroBottom {
  padding: 0 0 2px;
}

.GlobalTrendsListItem--marginZero {
  margin: 10px 0;
}

.GlobalTrendsListItem__expandedView {
  display: none;

  &.expanded {
    padding: 10px 25px;
    height: 250px;
    display: block;
  }
}

.GlobalTrendsListItem__leftSide {
  margin-right: 30px;
  height: 250px;
}

.GlobalTrendsListItem__noDataText {
  margin: 0;
  padding: 10px 0 0;
  color: $ghosty-blue;

  &.reduceTopPadding {
    padding-top: 4px;
  }

  &.addBottomPadding {
    padding-bottom: 32px;
  }
}

.GlobalTrendsListItem__lineGraph {
  margin-bottom: 30px;
  width: 250px;
  height: 60px;
}

.GlobalTrendsListItem__graphGroup {
  padding-left: 12px;

  > p {
    margin: 0 0 5px;
  }
}
