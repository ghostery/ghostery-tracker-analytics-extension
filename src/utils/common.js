/**
 * Common Methods
 *
 * Methods used in both /src and /apps
 *
 * Insights by Ghostery
 */

import Fuse from 'fuse.js';
import Globals from '../classes/Globals';

export function log(...args) {
  if (!Globals.LOG) { return; }

  args.unshift(`${(new Date()).toLocaleTimeString()}\t`);
  const error = /error/gi.test(args.toString());
  if (error) {
    console.error(...args); // eslint-disable-line no-console
  } else {
    console.log(...args); // eslint-disable-line no-console
  }
}

// TODO revise function description
// Determines the fuzzy search match value between
// a query string and a set of pattern strings
// params:
//  query: (required) string
//  patterns: (required) object array
//  keys: (required) string array. the key(s) in the patterns array objects to match against
//  options: (optional) object. Two options are supported:
//    isCaseSensitive: Boolean; default: false
//    threshold: Number between 0.0 and 1.0. 0.0 will include only perfect matches; 1.0 will match anything. default: 1.0
// returns: an array of { "pattern": [pattern String], "score": [Number between 0 and 1] }, sorted by score,
// and including all or a subset of the patterns depending on threshold value
export function fuzzySearch(query, patterns, keys, options = {}) {
  const fuseOptions = {
    keys,
    isCaseSensitive: options.isCaseSensitive || false,
    threshold: options.threshold || 0.2,
  };

  const isPatternsAnArray = Array.isArray(patterns);
  let patternsArray;
  if (isPatternsAnArray) {
    patternsArray = patterns;
  } else {
    patternsArray = Object.entries(patterns).map(([key, value]) => ({
      ...value,
      __fuzzySearchKey: key,
    }));
  }

  const fuse = new Fuse(patternsArray, fuseOptions);
  const fuseResults = fuse.search(query);
  const results = fuseResults.map(fuseResult => (
    fuseResult.item
  ));

  if (isPatternsAnArray) {
    return results;
  }

  const resultsObject = {};
  results.forEach((result) => {
    const key = result.__fuzzySearchKey;
    delete result.__fuzzySearchKey;
    resultsObject[key] = result;
  });
  return resultsObject;
}

// ToDo: Remove when we have another function export from this file.
export function test() {}
