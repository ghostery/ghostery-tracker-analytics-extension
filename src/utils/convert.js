/**
 * String and Number Utilities for Readability
 *
 * Insights by Ghostery
 */

const truncateNum = num => Math.round(num * 100) / 100;

// NOTE: Content-Length header field is in bytes (ie. octets)
export const convertBytes = (bytes) => {
  if (bytes > 1000000000) { return `${truncateNum(bytes / 1000000000)} GB`; }
  if (bytes > 1000000) { return `${truncateNum(bytes / 1000000)} MB`; }
  if (bytes > 1000) { return `${truncateNum(bytes / 1000)} KB`; }
  return `${truncateNum(bytes)} B`;
};

export const convertMilliseconds = (ms) => {
  if (ms > 3600000) { return `${truncateNum(ms / 3600000)} hours`; }
  if (ms > 60000) { return `${truncateNum(ms / 60000)} min`; }
  if (ms > 1000) { return `${truncateNum(ms / 1000)} sec`; }
  return `${truncateNum(ms)} ms`;
};

export const truncateString = (str, maxLength = 30) => {
  const maxNotCountingEllipsis = maxLength - 3;
  if (maxNotCountingEllipsis <= 0) { return str.slice(0, maxLength); }
  if (str.length > maxLength) { return (`${str.slice(0, maxNotCountingEllipsis)}...`); }
  return str;
};
