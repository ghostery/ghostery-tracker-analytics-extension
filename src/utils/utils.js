/**
 * Utilities
 *
 * Methods and properties that are used ONLY by src modules.
 *
 * Insights by Ghostery
 */

import debounce from 'lodash.debounce';

import Url from 'url';
import { log } from './common';

export function isValidTopLevelNavigation(details) {
  const {
    url,
    tabId,
    frameId,
  } = details;

  return frameId === 0
    && tabId > 0
    && url.startsWith('http')
    && !url.startsWith('https://chrome.google.com/webstore/');
}

export function isValidUrl(url) {
  if (url.protocol.startsWith('http')
    && url.host.includes('.')
    && /[A-Za-z]/.test(url.host)
    && !url.path.includes('_/chrome/newtab')) {
    return true;
  }

  return false;
}

export function processUrl(src) {
  const {
    protocol,
    hostname,
    pathname,
    host,
    hash,
  } = src ? Url.parse(src) : {};

  return {
    protocol: protocol ? protocol.substr(0, protocol.length - 1) : '',
    host: hostname || '',
    path: pathname ? pathname.substr(1) : '',
    hostWithPath: (host || '') + (pathname || ''),
    anchor: hash ? hash.substr(1) : '',
  };
}

/**
 * Process URLs and returns the query string as an object.
 * @memberOf BackgroundUtils
 * @param  {string} src   the source url
 * @return {Object}       contains parts of parsed query as its properties
 */
export function processUrlQuery(src) {
  if (!src) {
    return {};
  }

  try {
    const res = {};
    for (const [key, value] of new URL(src).searchParams.entries()) {
      res[key] = value;
    }
    return res;
  } catch (e) {
    return {};
  }
}

export function getQueryParam(src, param) {
  const { search } = new URL(src);
  const queryParams = new URLSearchParams(search);
  return queryParams.get(param);
}

// ToDo: Review these functions more closely.
function _fetchJson(method, url, query, extraHeaders, referrer = 'no-referrer', credentials = 'omit') {
  if (typeof fetch === 'function') {
    const headers = new Headers({
      'Content-Type': 'application/json',
      Accept: 'application/json',
    });
    // if (extraHeaders) {
    //   for (const [key, value] of objectEntries(extraHeaders)) {
    //     headers.append(key, value);
    //   }
    // }
    const options = {
      method,
      headers,
      body: query,
      referrerPolicy: referrer,
      credentials,
    };
    if (method === 'GET' || method === 'HEAD') {
      // Edge fails to construct Request object for GET and HEAD methods when body is present
      delete options.body;
    }

    const request = new Request(url, options);
    return fetch(request).then((response) => {
      const contentType = response.headers.get('content-type');
      if (!response.ok) {
        return Promise.reject(new Error(`Failed to fetch ${url} with status ${response.status} (${response.statusText})`));
      }
      // check for 204 status (No Content) from CMP
      if (response.status === 204) {
        return false; // send back false to signal no new campaigns
      }
      if (contentType && contentType.includes('application/json')) {
        return response.json();
      }
      if (contentType && contentType.includes('text/html')) {
        return response.text();
      }
      return response.text();
    }).then((data) => {
      // GET /api/Sync/ returns json with content-type:text/html. go figure.
      if (typeof data === 'string' && data.includes('{')) {
        try {
          log('_fetchJson resolved', (data) ? JSON.parse(data) : {});
          // attempt to parse the response.text as json
          return (data) ? JSON.parse(data) : {};
        } catch (err) {
          log('_fetchJson error', err);
          return Promise.reject(new Error(err));
        }
      } else {
        // data is either false or actual application/json
        return data;
      }
    }).catch((err) => {
      log(`_fetchJson Error: ${err}`);
      return Promise.reject(new Error(err));
    });
  }
  return new Promise(((resolve, reject) => {
    const xhr = new XMLHttpRequest();

    xhr.onload = () => {
      // This is called even on 404 etc, so check the status.
      if (xhr.status >= 200 && xhr.status < 400) {
        // check for 204 status (No Content) from CMP
        if (xhr.status === 204) {
          resolve(false); // send back false to signal no new campaigns
        } else if (xhr.responseText.includes('{')) {
          // For saveUserSettings we only get back a string that cannot be parsed
          try {
            log('_fetchJson resolved', (xhr.responseText) ? JSON.parse(xhr.responseText) : {});
            // Resolve the promise with the response text
            resolve((xhr.responseText) ? JSON.parse(xhr.responseText) : {});
          } catch (err) {
            log('_fetchJson error', err);
            reject(new Error(err));
          }
        } else {
          resolve(xhr.responseText);
        }
      } else {
        // Otherwise reject with the status text
        log('_fetchJson error', xhr.statusText);
        reject(new Error(xhr.statusText));
      }
    };

    // Handle network errors
    xhr.onerror = (error) => {
      log('_fetchJson network error', error);
      reject(new Error(error));
    };

    // Make the request
    log('_fetchJson request', method, url, query, extraHeaders);
    xhr.open(method, url, true);
    xhr.setRequestHeader('Content-Type', 'application/json');
    xhr.setRequestHeader('Accept', 'application/json');
    // if (extraHeaders) {
    //   for (const [key, value] of objectEntries(extraHeaders)) {
    //     xhr.setRequestHeader(key, value);
    //   }
    // }
    xhr.overrideMimeType('application/json');
    xhr.send(query);
  }));
}

export function getJson(url, extraHeaders) {
  return _fetchJson('GET', url, null, extraHeaders).catch((error) => {
    log('getJson error', error);
    return Promise.reject(error);
  });
}

export function fetchLocalJSONResource(url) {
  if (typeof fetch === 'function') {
    return fetch(url).then((response) => {
      if (!response.ok) {
        return Promise.reject(new Error(`Failed to fetchLocalJSONResource ${url} with status ${response.status} (${response.statusText})`));
      }
      return response.json();
    }).catch((err) => {
      log(`fetchLocalJSONResource error: ${err}`);
      return Promise.reject(new Error(err));
    });
  }
  return new Promise(((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.onload = () => {
      // This is called even on 404 etc, so check the status.
      if (xhr.status >= 200 && xhr.status < 400) {
        try {
          resolve((xhr.responseText) ? JSON.parse(xhr.responseText) : {});
        } catch (err) {
          log('fetchLocalJSONResource error', err);
          reject(new Error(err));
        }
      } else {
        // Otherwise reject with the status text
        log('fetchLocalJSONResource error', xhr.statusText);
        reject(new Error(xhr.statusText));
      }
    };

    // Handle network errors
    xhr.onerror = (error) => {
      log('fetchLocalJSONResource network error', error);
      reject(new Error(error));
    };

    // Make the request
    log('fetchLocalJSONResource request', url);
    xhr.open('GET', url, true);
    xhr.overrideMimeType('image/png');
    xhr.send();
  }));
}

export function defineLazyProperty(obj, prop, callback) {
  let value;
  let isSet = false;

  Object.defineProperty(obj, prop, {
    get() {
      if (!isSet) {
        value = callback();
        isSet = true;
      }
      return value;
    },
    set(val) {
      value = val;
      isSet = true;
    },
  });
}

export const flushChromeMemoryCache = debounce(() => {
  chrome.webRequest.handlerBehaviorChanged();
}, 1000 * 35, true);
