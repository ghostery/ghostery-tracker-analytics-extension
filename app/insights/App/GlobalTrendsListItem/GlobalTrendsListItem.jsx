import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Tooltip from 'react-bootstrap/Tooltip';

import GlobalTrendsLineGraph from '../GlobalTrendsLineGraph';
import GlobalTrendsSiteTypeGraph from '../GlobalTrendsSiteTypeGraph';
import categoryMapping from '../../../../src/utils/stringMaps';
import { truncateString } from '../../../../src/utils/convert';
import './GlobalTrendsListItem.scss';


class GlobalTrendsListItem extends Component {
  constructor(props) {
    super(props);

    this.state = { displayExpandedView: false };
  }

  handleClick = () => {
    const { displayExpandedView } = this.state;
    const { messageCreators } = this.props;
    if (!displayExpandedView) {
      messageCreators.sendMetrics({ type: 'tracker_expand_trackerInformationList' });
    }
    this.setState({ displayExpandedView: !displayExpandedView });
  }

  prepareDataForGraphs = () => {
    const { listItemWtmData } = this.props;
    if (!listItemWtmData) {
      return { reach: null, siteReach: null, categoryPresence: null, lastUpdated: null };
    }

    const siteReachLength = listItemWtmData.reach_time_series.site.length;
    const tsLength = listItemWtmData.reach_time_series.ts.length;
    const reach = [];
    const siteReach = [];
    const categoryPresence = [];
    listItemWtmData.reach_time_series.ts.forEach((month, index) => {
      reach.push({ date: month, percentage: listItemWtmData.reach_time_series.page[index] });
      if (index > tsLength - siteReachLength - 1) {
        siteReach.push({
          date: month,
          percentage: listItemWtmData.reach_time_series.site[tsLength - index - 1] / 10000,
        });
      }
    });

    listItemWtmData.presence_by_category.forEach((presence) => {
      categoryPresence.push({ name: presence[0], percentage: presence[1] / 100 });
    });

    reach.sort((a, b) => (new Date(a.date) - new Date(b.date)));

    const date = new Date(reach[reach.length - 1].date);
    date.setDate(date.getDate() + 1);
    const dateString = date.toDateString();
    const lastUpdated = `${dateString.slice(4, 8)}${dateString.slice(11, 15)}`;

    return { reach, siteReach, categoryPresence, lastUpdated };
  }

  render() {
    const {
      id, listItemWtmData, handleStar, starred, name, category,
    } = this.props;
    const { displayExpandedView } = this.state;
    const expandedView = ClassNames('GlobalTrendsListItem__expandedView', {
      expanded: displayExpandedView,
    });
    const GlobalTrendsListItemCircle = ClassNames(
      'GlobalTrendsListItem__circle',
      category,
    );
    const starIconFilePath = starred
      ? 'dist/images/shared/star-filled.svg'
      : 'dist/images/shared/star-empty.svg';

    const { reach, siteReach, categoryPresence, lastUpdated } = this.prepareDataForGraphs();

    return (
      <div className="GlobalTrendsListItem">
        <div
          className="GlobalTrendsListItem__header d-flex justify-content-between"
          onClick={this.handleClick}
        >
          <Container fluid>
            <Row className="GlobalTrendsListItem--paddingZeroTop">
              <Col className="GlobalTrendsListItem--paddingZeroBottom d-flex align-items-center" sm={1}>
                <img
                  alt="Star Tracker"
                  src={[chrome.extension.getURL(starIconFilePath)]}
                  className="GlobalTrendsListItem__star GlobalTrendsListItem--paddingZero"
                  onClick={(e) => {
                    e.stopPropagation();
                    handleStar(id, name);
                  }}
                />
              </Col>
              <Col
                className="GlobalTrendsListItem--paddingZero d-flex align-items-center justify-content-center"
                sm={1}
              >
                {listItemWtmData ? listItemWtmData.overview.reach_rank : (
                  <img
                    alt="No Data For This Tracker"
                    src="/dist/images/app/small-bar.svg"
                    width="15"
                    height="3"
                  />
                )}
              </Col>
              <Col className="GlobalTrendsListItem--paddingZero" sm={1} />
              <Col className="GlobalTrendsListItem--paddingZero" sm={5}>
                {truncateString(name, 40)}
              </Col>
              <Col className="GlobalTrendsListItem--paddingZero d-flex flex-row align-items-center" sm={4}>
                <div className={GlobalTrendsListItemCircle} />
                <div className="GlobalTrendsListItem--paddingZero">
                  {categoryMapping[category]}
                </div>
              </Col>
            </Row>
          </Container>
        </div>
        <div className={expandedView}>
          <span className="d-flex">
            <span className="GlobalTrendsListItem__leftSide d-flex flex-column justify-content-start">
              <p>Tracker Reach</p>
              <div className="GlobalTrendsListItem__lineGraph d-flex justify-content-start align-items-center">
                <img
                  src="/dist/images/app/tracker-info-traffic.svg"
                  width="32"
                  height="33"
                  alt="Tracker info line graph (traffic)"
                />
                <span className="GlobalTrendsListItem__graphGroup d-flex flex-column justify-content-between">
                  <p className="GlobalTrendsListItem--smallFont">
                    {`Web traffic that is tracked by ${name}`}
                  </p>
                  {listItemWtmData && displayExpandedView ? (
                    <span className="d-flex justify-content-start align-items-center">
                      <p className="GlobalTrendsListItem--largeFont">
                        {`${(listItemWtmData.overview.reach * 100).toFixed(2)} %`}
                      </p>
                      <GlobalTrendsLineGraph data={reach} />
                    </span>
                  ) : (
                    <p className="GlobalTrendsListItem__noDataText">
                      Data not yet available
                    </p>
                  )}
                </span>
              </div>
              <div className="GlobalTrendsListItem__lineGraph d-flex justify-content-start align-items-center">
                <img
                  src="/dist/images/app/tracker-info-site.svg"
                  width="32"
                  height="33"
                  alt="Tracker info line graph (sites)"
                />
                <span className="GlobalTrendsListItem__graphGroup d-flex flex-column justify-content-between">
                  <p className="GlobalTrendsListItem--smallFont">
                    {`Sites using the ${name} tracker`}
                  </p>
                  {listItemWtmData && displayExpandedView ? (
                    <span className="d-flex justify-content-start align-items-center">
                      <p className="GlobalTrendsListItem--largeFont">
                        {`${listItemWtmData.overview.site_reach_top10k / 100} %`}
                      </p>
                      <GlobalTrendsLineGraph data={siteReach} />
                    </span>
                  ) : (
                    <p className="GlobalTrendsListItem__noDataText">
                      Data not yet available
                    </p>
                  )}
                </span>
              </div>
              <p className="GlobalTrendsListItem--smallFont">
                {listItemWtmData && `Data last updated in ${lastUpdated}`}
              </p>
            </span>
            <span className="GlobalTrendsListItem__rightSide d-flex flex-column justify-content-start">
              <p>Seen On These Types Of Websites</p>
              {listItemWtmData && displayExpandedView ? (
                <GlobalTrendsSiteTypeGraph data={categoryPresence} />
              ) : (
                <p className="GlobalTrendsListItem__noDataText reduceTopPadding addBottomPadding">
                  Data not yet available
                </p>
              )}
              <p className="GlobalTrendsListItem--marginZero">Tracking Methods</p>
              {!listItemWtmData && (
                <p className="GlobalTrendsListItem__noDataText reduceTopPadding">
                  Data not yet available
                </p>
              )}
              {listItemWtmData && listItemWtmData.tracking_method.cookies && (
                <span className="GlobalTrendsListItem__trackingType d-flex justify-content-start align-items-center">
                  <img
                    src="/dist/images/app/cookies.svg"
                    width="24"
                    height="24"
                    alt="Cookie icon (tracking type)"
                  />
                  <p className="GlobalTrendsListItem__trackingTypeText">Cookies</p>
                  <OverlayTrigger
                    placement="right"
                    overlay={(
                      <Tooltip className="tooltipLeft">
                        Files placed by a website into the browser that are used to identify the user.
                      </Tooltip>
                    )}
                  >
                    <img
                      src="/dist/images/shared/info-icon.svg"
                      className="infoIcon"
                      width="16"
                      height="16"
                      alt="More info hover"
                    />
                  </OverlayTrigger>
                </span>
              )}
              {listItemWtmData && listItemWtmData.tracking_method.fingerprinting && (
                <span className="GlobalTrendsListItem__trackingType d-flex justify-content-start align-items-center">
                  <img
                    src="/dist/images/app/fingerprint.svg"
                    width="20"
                    height="23"
                    alt="Fingerprinting icon (tracking type)"
                  />
                  <p className="GlobalTrendsListItem__trackingTypeText">
                    Fingerprinting
                  </p>
                  <OverlayTrigger
                    placement="right"
                    overlay={(
                      <Tooltip className="tooltipLeft">
                        Data collected for the purpose of identification and can be used to identify individuals or devices even when cookies are turned off
                      </Tooltip>
                    )}
                  >
                    <img
                      src="/dist/images/shared/info-icon.svg"
                      className="infoIcon"
                      width="16"
                      height="16"
                      alt="More info hover"
                    />
                  </OverlayTrigger>
                </span>
              )}
              {listItemWtmData && !listItemWtmData.tracking_method.cookies
              && !listItemWtmData.tracking_method.fingerprinting && (
                <p className="GlobalTrendsListItem__notDetectedText">
                  Method not detected
                </p>
              )}
            </span>
          </span>
        </div>
      </div>
    );
  }
}

GlobalTrendsListItem.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  messageCreators: PropTypes.shape({
    sendMetrics: PropTypes.func.isRequired,
  }).isRequired,
  handleStar: PropTypes.func.isRequired,
  starred: PropTypes.bool.isRequired,
  listItemWtmData: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.shape({
      tracking_method: PropTypes.shape({
        cookies: PropTypes.bool.isRequired,
        fingerprinting: PropTypes.bool.isRequired,
      }).isRequired,
      presence_by_category: PropTypes.arrayOf(PropTypes.arrayOf(
        PropTypes.oneOfType([
          PropTypes.string.isRequired,
          PropTypes.number.isRequired,
        ]).isRequired,
      )).isRequired,
      overview: PropTypes.shape({
        reach: PropTypes.number.isRequired,
        site_reach_top10k: PropTypes.number.isRequired,
        reach_rank: PropTypes.number.isRequired,
      }).isRequired,
      reach_time_series: PropTypes.shape({
        page: PropTypes.arrayOf(PropTypes.number).isRequired,
        ts: PropTypes.arrayOf(PropTypes.string).isRequired,
        site: PropTypes.arrayOf(PropTypes.number).isRequired,
      }).isRequired,
    }),
  ]).isRequired,
};

export default GlobalTrendsListItem;
