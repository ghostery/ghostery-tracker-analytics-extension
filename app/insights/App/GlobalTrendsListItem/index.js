import { connect } from 'react-redux';

import GlobalTrendsListItem from './GlobalTrendsListItem.jsx';
import {
  sendMetrics,
} from '../../../store/messages/messageCreators';

const mapStateToProps = () => ({
  messageCreators: { sendMetrics },
});

export default connect(mapStateToProps)(GlobalTrendsListItem);
