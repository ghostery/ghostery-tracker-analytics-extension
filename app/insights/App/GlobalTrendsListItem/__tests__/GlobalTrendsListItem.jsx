import React from 'react';
import { shallow } from 'enzyme';
import chrome from 'sinon-chrome';
import * as dependency from '../../../../../src/utils/convert';
import GlobalTrendsListItem from '../GlobalTrendsListItem';

jest.mock('../../../../../src/classes/Globals', () => ({
  BROWSER_INFO: { displayName: 'Chrome' },
}));

describe('app/insights/App/insights/GlobalTrendsListItem', () => {
  describe('Snapshot tests with react-test-renderer', () => {
    const globalTrendsListItem = {
      id: 'testing',
      name: 'name-testing',
      category: 'advertising-testing',
      messageCreators: {
        sendMetrics: () => {},
      },
      handleStar: () => {},
      starred: true,
      listItemWtmData: false,
    };

    test('GlobalTrendsListItem component should render', () => {
      global.chrome = chrome;

      dependency.truncateString = jest.fn();

      const component = shallow(
        <GlobalTrendsListItem
          {...globalTrendsListItem}
        />,
      );
      expect(component).toMatchSnapshot();
    });

    test('GlobalTrendsListItem handleClick should resolve', () => {
      let displayExpandedView;
      const component = shallow(
        <GlobalTrendsListItem
          {...globalTrendsListItem}
        />,
      );
      component.setState({ displayExpandedView: !displayExpandedView });
      expect(component.length).toBe(1);
    });

    test('GlobalTrendsListItem sendMetrics should resolve', () => {
      global.chrome = chrome;
      dependency.truncateString = jest.fn();

      const component = shallow(<GlobalTrendsListItem {...globalTrendsListItem} />);
      component.setState({ type: 'tracker_expand_trackerInformationList' });
      component.find('div.GlobalTrendsListItem__header').simulate('click');
      expect(component.find('div.GlobalTrendsListItem__header').length).toBe(1);
    });
  });
});
