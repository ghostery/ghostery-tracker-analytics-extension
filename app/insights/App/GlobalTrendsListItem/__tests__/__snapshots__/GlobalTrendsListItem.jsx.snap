// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/insights/App/insights/GlobalTrendsListItem Snapshot tests with react-test-renderer GlobalTrendsListItem component should render 1`] = `
<div
  className="GlobalTrendsListItem"
>
  <div
    className="GlobalTrendsListItem__header d-flex justify-content-between"
    onClick={[Function]}
  >
    <Container
      fluid={true}
    >
      <ForwardRef
        className="GlobalTrendsListItem--paddingZeroTop"
        noGutters={false}
      >
        <Col
          className="GlobalTrendsListItem--paddingZeroBottom d-flex align-items-center"
          sm={1}
        >
          <img
            alt="Star Tracker"
            className="GlobalTrendsListItem__star GlobalTrendsListItem--paddingZero"
            onClick={[Function]}
            src={
              Array [
                undefined,
              ]
            }
          />
        </Col>
        <Col
          className="GlobalTrendsListItem--paddingZero d-flex align-items-center justify-content-center"
          sm={1}
        >
          <img
            alt="No Data For This Tracker"
            height="3"
            src="/dist/images/app/small-bar.svg"
            width="15"
          />
        </Col>
        <Col
          className="GlobalTrendsListItem--paddingZero"
          sm={1}
        />
        <Col
          className="GlobalTrendsListItem--paddingZero"
          sm={5}
        />
        <Col
          className="GlobalTrendsListItem--paddingZero d-flex flex-row align-items-center"
          sm={4}
        >
          <div
            className="GlobalTrendsListItem__circle advertising-testing"
          />
          <div
            className="GlobalTrendsListItem--paddingZero"
          />
        </Col>
      </ForwardRef>
    </Container>
  </div>
  <div
    className="GlobalTrendsListItem__expandedView"
  >
    <span
      className="d-flex"
    >
      <span
        className="GlobalTrendsListItem__leftSide d-flex flex-column justify-content-start"
      >
        <p>
          Tracker Reach
        </p>
        <div
          className="GlobalTrendsListItem__lineGraph d-flex justify-content-start align-items-center"
        >
          <img
            alt="Tracker info line graph (traffic)"
            height="33"
            src="/dist/images/app/tracker-info-traffic.svg"
            width="32"
          />
          <span
            className="GlobalTrendsListItem__graphGroup d-flex flex-column justify-content-between"
          >
            <p
              className="GlobalTrendsListItem--smallFont"
            >
              Web traffic that is tracked by name-testing
            </p>
            <p
              className="GlobalTrendsListItem__noDataText"
            >
              Data not yet available
            </p>
          </span>
        </div>
        <div
          className="GlobalTrendsListItem__lineGraph d-flex justify-content-start align-items-center"
        >
          <img
            alt="Tracker info line graph (sites)"
            height="33"
            src="/dist/images/app/tracker-info-site.svg"
            width="32"
          />
          <span
            className="GlobalTrendsListItem__graphGroup d-flex flex-column justify-content-between"
          >
            <p
              className="GlobalTrendsListItem--smallFont"
            >
              Sites using the name-testing tracker
            </p>
            <p
              className="GlobalTrendsListItem__noDataText"
            >
              Data not yet available
            </p>
          </span>
        </div>
        <p
          className="GlobalTrendsListItem--smallFont"
        />
      </span>
      <span
        className="GlobalTrendsListItem__rightSide d-flex flex-column justify-content-start"
      >
        <p>
          Seen On These Types Of Websites
        </p>
        <p
          className="GlobalTrendsListItem__noDataText reduceTopPadding addBottomPadding"
        >
          Data not yet available
        </p>
        <p
          className="GlobalTrendsListItem--marginZero"
        >
          Tracking Methods
        </p>
        <p
          className="GlobalTrendsListItem__noDataText reduceTopPadding"
        >
          Data not yet available
        </p>
      </span>
    </span>
  </div>
</div>
`;
